gc.c

```diff
@@ -2749,11 +2749,19 @@ rb_class_allocate_instance(VALUE klass)
     return obj;
 }
 
+static inline void
+rb_data_object_check(VALUE klass)
+{
+    if (klass != rb_cObject && (rb_get_alloc_func(klass) == rb_class_allocate_instance)) {
+        rb_warn("T_DATA class %"PRIsVALUE" should undefine or redefine the allocate method, please see doc/extension.rdoc", klass);
```

Comment:
Warn just once, and raise an exception if `new` or `allocate` are called later.

```suggestion
        rb_undef_alloc_func(klass);
#if 0 // TODO: enable at the next release
        rb_warn("allocator of T_DATA class %"PRIsVALUE" got undefined", klass);
#endif
```

Reply:
Thank you for this guidance, I've force-pushed with this change made.